### PR TITLE
Harsh/add doctor

### LINKED ISSUE #5: Add Doctor Screen
**Create screen for adding doctor using  [screen](https://drive.google.com/file/d/1CrsR4P_3LviwrtNSoOD5XPg0GT4gvAjs/view?usp=drive_link).** 

ACs

- [ ] Screen will have field **userId, name, email, address(ignore in UI), contactNumber, degree(ignore in UI), speciality(ignore in UI), availableDate(ignore in UI), availableTime(ignore in UI), images(ignore in UI), isNotificationEnable(toggle), isEmailVerified(ignore in UI, pass value as false, ), isPhoneNumverVerifiedd(ignore in UI, pass value as false), isDoctor(ignore in UI pass value as true), isAdmin(ignore in UI pass value as false)**
- [x] Each fields  which  are in UI aremandatory so it should have validation, submit button will enabled after user enter valid data in all fields.
- [x] Name validation - Should contain at least 3 character. validation message "Name should contain at least 3 characters" 
- [x] Email - no null And valid email
- [x] On submit click data should save in firestore table name **user_data**
- [x] After data added successfully it should display "Doctor information successfully saved."
- [x] User should move to login screen

### Check List

- [x] I have linked my PR with this task
- [x] I have added comment after completing this task
- [x] I have followed all [guidelines](https://github.com/jaykishan-sewak/doctor-appoint-app/tree/main) for code

### PR DESCRIPTION
**Pull request by HarshPanchalInfusion**
Completed Work on add doctor screen, with all validation & data store in firedtore.

### CODE REVIEW

File: app/src/main/java/com/android/doctorapp/repository/AddDoctorRepository.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import javax.inject.Inject
+
+class AddDoctorRepository @Inject constructor() {
+
+    suspend fun addDoctorData(doctorRequestModel: UserDataRequestModel, firestore: FirebaseFirestore): ApiResponse<StringResponse> {
+        return try {
+            val addDoctorResponse = firestore.collection("user_data").add(doctorRequestModel).await()
+            ApiResponse.create(response = Response.success(StringResponse("Success")))
+        } catch (e: Exception) {
+            ApiResponse.create(response = Response.success(StringResponse(e.message.toString())))
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/repository/AddDoctorRepository.kt]: Incorrect code

ApiResponse.create(response = Response.success(StringResponse(e.message.toString())))

File: app/src/main/java/com/android/doctorapp/repository/models/StringResponse.kt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package com.android.doctorapp.repository.models
+
+data class StringResponse(val message: String)
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/repository/models/StringResponse.kt]: Add Doctor response. Please update file name

File: app/src/main/java/com/android/doctorapp/repository/network/AppApi.kt [hunk 3]
```diff
@@ -28,4 +29,9 @@ interface AppApi {
     suspend fun register(
         @Body registerRequestModel: RegisterRequestModel
     ): Response<LoginResponseModel>
+
+    @POST(END_URL_ADD_DOCTOR_DATA)
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/repository/network/AppApi.kt]: No need to mention @Post and end point url

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                    resourceProvider.getString(R.string.error_valid_phone_number)
+            }
+        }
+        isAllValidate()
+    }
+
+
+    fun addDoctorData() {
+        // Initialize firebase auth
+        firebaseAuth = FirebaseAuth.getInstance()
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Please refer auth code for login screen

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        val firestore = FirebaseFirestore.getInstance()
+
+        // Initialize firebase user
+        val firebaseUser = firebaseAuth.currentUser
+        if (firebaseUser != null) {
+            // when firebaseUser is not null then
+            val userData = UserDataRequestModel(
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt:100]: User default argument in data class. After that no need to define each and every field

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    private val _navigationListener = SingleLiveEvent<Int>()
+    val navigationListener = _navigationListener.asLiveData()
+
+    private val _addDoctorResponse = SingleLiveEvent<String>()
+    val addDoctorResponse = _addDoctorResponse.asLiveData()
+
+    private lateinit var fireStore: FirebaseFirestore
+    private lateinit var firebaseUser: FirebaseUser
+
+    init {
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Please define all primary type object define  in baseViewModel. So any View Model can access those objects.

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+        firebaseUser = auth?.currentUser!!
+        if (firebaseUser != null) {
+            // when firebaseUser is not null then
+            val userData = UserDataRequestModel(
+                userId = firebaseUser.uid,
+                isDoctor = true,
+                email = doctorEmail.value,
+                name = doctorName.value,
+                contactNumber = doctorContactNumber.value,
```
Comment (vikas-infusion) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Here only pass form data that is show in UI. Still passing other data. Put default value on data class.

not define each and every field default value null.

File: app/src/main/res/layout/fragment_add_doctor.xml [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            android:layout_height="wrap_content"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
+
+        <ScrollView
+            android:layout_width="match_parent"
+            android:layout_height="0dp"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
```
Comment (vikas-infusion) [inline app/src/main/res/layout/fragment_add_doctor.xml:32]: Put horizontal and vertical space here for all fields that coming in UI.

File: app/src/main/res/layout/fragment_add_doctor.xml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@+id/ic_add_doctor">
+
+                    <com.google.android.material.textfield.TextInputEditText
+                        android:id="@+id/et_doctor_name"
+                        android:layout_width="match_parent"
+                        android:layout_height="wrap_content"
+                        android:afterTextChanged="@{(text) -> viewModel.isValidateName(text)}"
+                        android:hint="@string/hint_doctor_name"
+                        android:imeOptions="actionNext"
```
Comment (vikas-infusion) [inline app/src/main/res/layout/fragment_add_doctor.xml:70]: Please make common style for input box. 

Like textColor,fontSize,singleLine etc.

File: app/src/main/res/layout/fragment_add_doctor.xml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+                        android:inputType="textEmailAddress"
+                        android:maxLines="1"
+                        android:text="@={viewModel.doctorEmail}"
+                        android:textColor="@color/black"
+                        android:textSize="14sp" />
+
+                </com.google.android.material.textfield.TextInputLayout>
+
+                <!--                        android:onTextChanged="@{(text, start, before, count) -> viewModel.isValidEmail(text)}"-->
```
Comment (vikas-infusion) [inline app/src/main/res/layout/fragment_add_doctor.xml]: Remove commented code

File: app/src/main/res/layout/fragment_add_doctor.xml [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/margin_20"
+                    android:layout_marginTop="@dimen/margin_10"
+                    android:text="@string/doctor_notification_enable"
+                    android:textColor="@color/black"
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@id/field_doctor_contact" />
+
+                <androidx.appcompat.widget.AppCompatToggleButton
```
Comment (vikas-infusion) [inline app/src/main/res/layout/fragment_add_doctor.xml]: please use SwitchCompat control

File: app/src/main/res/layout/fragment_add_doctor.xml [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                <androidx.appcompat.widget.AppCompatButton
+                    android:id="@+id/button_submit"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/margin_10"
+                    android:enabled="@{viewModel.isDataValid}"
+                    android:onClick="@{()->viewModel.addDoctorData()}"
+                    android:text="@string/submit"
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintStart_toStartOf="parent"
```
Comment (vikas-infusion) [inline app/src/main/res/layout/fragment_add_doctor.xml:155]: if you are using layout_constraintStart_toStartOf then please define layout_width 0dp

```
    android:layout_width="0dp"
        app:layout_constraintEnd_toEndOf="parent"
                    app:layout_constraintStart_toStartOf="parent"
```

**Review by vikas-infusion — DISMISSED** [review]

File: app/src/main/java/com/android/doctorapp/repository/AddDoctorRepository.kt (hunk 1 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/repository/AddDoctorRepository.kt]: Done

File: app/src/main/java/com/android/doctorapp/repository/models/StringResponse.kt (hunk 2 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/repository/models/StringResponse.kt]: Done

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt (hunk 4 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Done

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt (hunk 6 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Done

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt (hunk 7 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Done

File: app/src/main/res/layout/fragment_add_doctor.xml (hunk 11 above)
Comment (HarshPanchalInfusion) [reply app/src/main/res/layout/fragment_add_doctor.xml]: Done

File: app/src/main/res/layout/fragment_add_doctor.xml (hunk 10 above)
Comment (HarshPanchalInfusion) [reply app/src/main/res/layout/fragment_add_doctor.xml]: Done

File: app/src/main/java/com/android/doctorapp/repository/network/AppApi.kt (hunk 3 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/repository/network/AppApi.kt]: Done

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    val isDataValid: MutableLiveData<Boolean> = MutableLiveData(false)
+
+    private val _navigationListener = SingleLiveEvent<Int>()
+    val navigationListener = _navigationListener.asLiveData()
+
+    private val _addDoctorResponse = SingleLiveEvent<String>()
+    val addDoctorResponse = _addDoctorResponse.asLiveData()
+
+
+    private fun isAllValidate() {
```
Comment (jaykishan-sewak) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: rename to validateAllField

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        isDataValid.value = (!doctorName.value.isNullOrEmpty() && !doctorEmail.value.isNullOrEmpty()
+                && !doctorContactNumber.value.isNullOrEmpty() && doctorNameError.value.isNullOrEmpty()
+                && doctorEmailError.value.isNullOrEmpty() && doctorContactNumberError.value.isNullOrEmpty())
+    }
+
+    fun isValidateName(text: CharSequence?) {
```
Comment (jaykishan-sewak) [inline app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: remane to isValidName

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt (hunk 13 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Done

File: app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt (hunk 14 above)
Comment (HarshPanchalInfusion) [reply app/src/main/java/com/android/doctorapp/ui/doctor/AddDoctorViewModel.kt]: Dobe

**Review by vikas-infusion — APPROVED** [review]

**Pull request merged**